File: 程序员面试金典/8.4/subsets_2.py
# 方法2：递归
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        return get_subsets(nums, len(nums) - 1)


def get_subsets(nums, index):
    if index == -1:
        return [[]]
    else:
        subsets = get_subsets(nums, index - 1)
        return subsets + [subset + [nums[index]] for subset in subsets]

File: 程序员面试金典/8.4/subsets_2.java
// 方法2：递归
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        return getSubsets(nums, nums.length - 1);
    }

    private List<List<Integer>> getSubsets(int[] nums, int index) {
        List<List<Integer>> subsets;
        if (index == -1) {
            subsets = new ArrayList<>();
            subsets.add(new ArrayList<>());
        }
        else {
            subsets = getSubsets(nums, index - 1);
            int size = subsets.size();
            for (int i = 0; i < size; ++i) {
                List<Integer> temp = (ArrayList<Integer>) ((ArrayList<Integer>) subsets.get(i)).clone();
                temp.add(nums[index]);
                subsets.add(temp);
            }
        }
        return subsets;
    }

}
